### PR TITLE
Expose membership type in the Admin UI for organization members and users

### LINKED ISSUE #33498: Expose membership type in the Admin UI for organization members
### Description

For the sake of UX, we should expose the membership type when fetching the representation for users that belong to an organization. It should make it a lot easier to identify whether a user is a `managed` or `unmanaged` organization member and make it more clear how certain operations will impact the user. For instance, unlinking a `managed` member from an organization will remove the account from the realm whereas the account for an `unmanaged` member is kept and only the link to the organization is removed.

The Organization Admin API is already exposing the membership type of a member from the `/admin/realms/{realm}/organizations/{orgid}/members` endpoint when querying members of an organization. There is a `membershipType` property within the member representation.

In the Admin UI, we can probably have a column in the `Members` list to show the membership type. We could also expand the idea and perhaps have some filter to show only `managed` or `unmanaged` members (or all of them).

I'm not sure though if we want or how we want to do that same when managing users through the User UI. 

### Discussion

_No response_

### Motivation

Improve UX in the Admin UI when managing organization members.

### Details

_No response_

### PR DESCRIPTION
**Pull request by agagancarczyk**
Closes https://github.com/keycloak/keycloak/issues/33498

I have this implemented in Organizations > Organization details > Members and Users > User details > Organizations

To test this, you'll need to have managed and unmanaged users in the organization.

### CODE REVIEW

**Review by pedroigor — DISMISSED** [review]
Nice. LGTM!

File: js/apps/admin-ui/src/user/Organizations.tsx [hunk 1]
```diff
@@ -25,30 +26,111 @@ import { toEditOrganization } from "../organizations/routes/EditOrganization";
...
+        await adminClient.organizations.memberOrganizations({ userId: id! });
+
+      const userOrganizationsWithMembershipTypes = await Promise.all(
+        userOrganizations.map(async (org) => {
+          const orgId = org.id;
+
+          const memberships: MembershipTypeRepresentation[] =
+            await adminClient.organizations.listMembers({
+              orgId: orgId!,
+            });
```
Comment (edewit) [inline js/apps/admin-ui/src/user/Organizations.tsx:96]: Is it not slow to query all users of an organisation?

File: js/apps/admin-ui/src/user/Organizations.tsx [hunk 2]
```diff
@@ -25,30 +26,111 @@ import { toEditOrganization } from "../organizations/routes/EditOrganization";
...
+          );
+
+          const membershipType = userMemberships.map((membership) => {
+            const formattedMembershipType = capitalizeFirstLetterFormatter()(
+              membership.membershipType,
+            );
+            return formattedMembershipType;
+          });
+
+          return { ...org, membershipType };
```
Comment (edewit) [inline js/apps/admin-ui/src/user/Organizations.tsx:109]: this seems duplicated, you could maybe put it into a separated file

**Review by edewit — CHANGES_REQUESTED** [review]

**Comment by pedroigor** [comment]
@agagancarczyk Updated the member listing page in organizations to use the new filter. However, there are two issues I'm not sure how to proceed:

* If you provide a search string and select a membership type, the search string is lost and query will only filter by the selected membership type
* If the selected membership type filter does not return any result, you will see the "empty page" asking the administrator to add members to the organization

File: server-spi/src/main/java/org/keycloak/organization/OrganizationProvider.java [hunk 3]
```diff
@@ -139,8 +142,32 @@ default Stream<OrganizationModel> getAllStream() {
...
+    default Stream<UserModel> getMembersStream(OrganizationModel organization, Map<String, String> filters, Boolean exact, Integer first, Integer max) {
+        var result = getMembersStream(organization, Optional.ofNullable(filters).orElse(Map.of()).get(UserModel.SEARCH), exact, first, max);
+        var membershipType = Optional.ofNullable(filters.get(MembershipType.NAME)).map(MembershipType::valueOf).orElse(null);
+
+        if (membershipType != null) {
+            return result.filter(userModel -> MembershipType.MANAGED.equals(membershipType) ? isManagedMember(organization, userModel) : !isManagedMember(organization, userModel));
+        }
+
+        return result;
+    }
```
Comment (sguilhen) [inline server-spi/src/main/java/org/keycloak/organization/OrganizationProvider.java:170]: Not sure I like these defaults - even though they are overridden, the default impls are calling each other leading possibly to a stack overflow?

**Review by sguilhen — COMMENTED** [review]
@pedroigor I haven't really reviewed the UI side of things, just the backend, and in general it looks good to me. Only thing I really don't like is those default impls in the provider delegating to each other

File: js/apps/admin-ui/src/organizations/Members.tsx [hunk 4]
```diff
@@ -37,19 +43,70 @@ export const Members = () => {
...
+          search,
+          membershipType,
+        });
+
+      const formattedMemberships = memberships.map((membership) => ({
+        ...membership,
+        membershipType: capitalizeFirstLetterFormatter()(
+          membership.membershipType,
+        ) as string,
+      }));
```
Comment (edewit) [inline js/apps/admin-ui/src/organizations/Members.tsx:102]: This is not needed, we just need to translate this key

File: server-spi/src/main/java/org/keycloak/organization/OrganizationProvider.java (hunk 3 above)
Comment (pedroigor) [reply server-spi/src/main/java/org/keycloak/organization/OrganizationProvider.java:165]: True, done.

File: server-spi/src/main/java/org/keycloak/organization/OrganizationProvider.java [hunk 5]
```diff
@@ -139,8 +142,32 @@ default Stream<OrganizationModel> getAllStream() {
      *
      * @param organization the organization
      * @return Stream of the members. Never returns {@code null}.
+     * @deprecated Use {@link #getMembersStream(OrganizationModel, Map, Boolean, Integer, Integer)} instead.
      */
-    Stream<UserModel> getMembersStream(OrganizationModel organization, String search, Boolean exact, Integer first, Integer max);
+    @Deprecated(forRemoval = true, since = "26")
+    default Stream<UserModel> getMembersStream(OrganizationModel organization, String search, Boolean exact, Integer first, Integer max) {
```
[the remark this answers is not in the document]
Comment (pedroigor) [reply server-spi/src/main/java/org/keycloak/organization/OrganizationProvider.java]: True, done.

**Review by sguilhen — APPROVED** [review]

**Review by edewit — DISMISSED** [review]

**Review by pedroigor — APPROVED** [review]

**Review by edewit — APPROVED** [review]

**Pull request merged**